pep-0701.rst

```diff
@@ -0,0 +1,367 @@
+   how f-strings can be nested, but the fact that string quotes cannot be reused
+   inside the expression component of f-strings made it impossible to nest
+   f-strings arbitrarily. In fact, this is the most nested-fstring that can be
+   written::
+
+    >>> f"""{f'''{f'{f"{1+1}"}'}'''}"""
+    '2'
+
+   As this PEP allows placing **any** valid Python expression inside the
+   expression component of the f-strings, it is now possible to reuse quotes and
+   therefore is possible to nest f-strings arbitrarily::
+
+    >>> f"{f"{f"{f"{f"{f"{1+1}"}"}"}"}"}"
+    '2'
+   
+   Although this is just a consequence of allowing arbitrary expressions, the
+   authors of this PEP do not believe that this is a fundamental benefit and we
+   have decided that the language specification will not explicitly mandate that
+   this nesting can be arbitrary. This is because allowing arbitrarily-deep
+   nesting imposes a lot of extra complexity to the lexer implementation
+   (particularly as lexer/parser pipelines need to allow "untokenizing" to
+   support the 'f-string debugging expressions' and this is especially taxing when
+   arbitrary nesting is allowed). Implementations are therefore free to impose a
+   limit on the nesting depth if they need to. Note that this is not an uncommon
+   situation, as the CPython implementation already imposes several limits all
+   over the place, including a limit on the nesting depth of parentheses and
+   brackets, a limit on the nesting of the blocks, a limit in the number of
+   branches in ``if`` statements, a limit on the number of expressions in
+   star-unpacking, etc.
+
+Specification
+=============
+
+The formal proposed PEG grammar specification for f-strings is (see :pep:`617`
+for details on the syntax):
+
+.. code-block:: peg
+
+    fstring
+        | FSTRING_START fstring_middle* FSTRING_END
+    fstring_middle
+        | fstring_replacement_field
+        | FSTRING_MIDDLE
+    fstring_replacement_field
+        | '{' (yield_expr | star_expressions) "="? [ "!" NAME ] [ ':' fstring_format_spec* ] '}'
+    fstring_format_spec:
+        | FSTRING_MIDDLE
+        | fstring_replacement_field
+
+This PEP leaves up to the implementation the level of f-string nesting allowed.
+This means that limiting nesting is **not part of the language specification**
+but also the language specification **doesn't mandate arbitrary nesting**. 
+
+Three new tokens are introduced:
+
+* ``FSTRING_START``: This token includes f-string character (``f``/``F``) and the open quote(s).
+* ``FSTRING_MIDDLE``: This token includes the text between the opening quote
+  and the first expression brace (``{``) and the text between two expression braces (``}`` and ``{``).
+* ``FSTRING_END``: This token includes everything after the last expression brace (or the whole literal part
+  if no expression exists) until the closing quote.
+
+These tokens are always string parts and they are semantically equivalent to the
```

Comment:
We probably want to rewrite this to mention that these token descriptions are only really an implementation detail for the interpreter and that is subjected to change.

Reply:
If it's an implementation detail, should the tokens be in the PEP at all?

Reply:
Yes, what’s an implementation detail is what the tokens contain, but the tokens themselves are part of the grammar

Reply:
I know it sounds a bit hocus pocus. Hopefully it will be more clear when i rewrite that

Reply:
@pablogsal should we merge this one as is, or await the rewrite? (I think we can proceed and do the next iterations on top of the repo itself to start early to discussions)

Reply:
Yeah let’s merge this and iterate in more PRs